config_v2.md

```diff
@@ -0,0 +1,370 @@
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+```bash
+python my_main.py with config_change1
+python my_main.py with config_change2
+```
+
+By default, when calling the potential_modification from the command line, 
+you should specify the `__name__` of the function. If you want to be able to say exactly which name you want, you can do:
+
+```python
+@sacred.potential_modification('pretty_name_here')
+def config_change2849459(config):
+    ...
+```
+
+```bash
+python my_main.py with pretty_name_here
+```
+
+
+### Example with config updates depending on each other
+
+
+`config_change3` depend on some parameters of `config_change1`.
+
+```python
+import sacred
+from sacred import Config
+
+from somewhere_else import load_dataset
+
+
+def config_change1(config):
+    config['dataset_config'] = dict(crop_size=(30, 30), random_flip=True)
+    config['dataset_size'] = 500
+    
+
+def config_change2(config):
+    config['dataset_config'] = dict(crop_size=(50, 50), random_flip=False)
+    config['dataset_size'] = 700
+    
+    
+def config_change3(config):
+    config_change1(config)
+    config['dataset_size'] = 1_000_000
+
+
+potential_modifs = [config_change1, config_change2, config_change3]
+configuration = Config(dict(batch_size=32, dataset_size=10_000, nb_epochs=50),
+                       potential_modifications=potential_modifs)
+
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs, dataset_config):
+    # main experiment here
+    my_dataset = load_dataset(dataset_size, **dataset_config)
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+
+### Example with descriptions of the config values:
+
+
+The `Config` object will remove the `ConfigValues` and keep only the value, for ease of access.
+It will put the descriptions somewhere else, in another attribute of the `Config` object. 
+Let's say `Config.values_descriptions`. 
+
+Notice the way the batch size value is accessed when getting the default dataset size. The ConfigValue has been removed in the constructor.
+
+```python
+import sacred
+from sacred import ConfigValue, Config
+
+
+configuration = Config(dict(
+    batch_size=ConfigValue(32, "The batch size value"),
+    dataset_size=ConfigValue(lambda config: config['batch_size'] * 100, "The dataset size", delayed=True),
+    nb_epochs=ConfigValue(50, "The number of epochs")
+))
+print(configuration['batch_size'])  # will print 32, not a ConfigValue.
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+
+### Example with dynamic population of the potential modifications.
+
+
+There is a hierarchy now. Similar to ingredients.
```

Comment:
I'd find this change a bit of regression in functionality for me: I quite enjoy the current ingredient design insofar as an Ingredient being itself an Experiment means that I can switch workflow easily between "debug this ingredient on its own by running it as an independent experiment" and "plug this same ingredient in to a larger macro-experiment". For example, if I had ingredients for "pull data" and "fit model" and "perform certain tests with a model", during development I might want to use cached versions of a fit model while I'm debugging the testing step, but then be easily able to chain the three ingredients together in an Experiment when I run them for real.

Reply:
Thanks a lot for describing your workflow. Let me make sure I understand. You are talking about testing individual parts of your experiment with the help of configs present in ingredients, right? Are you using any testing framework like Unittest or Pytest on top of it? We have to make sure the new API allow easy testing, so your use case have to be supported.

Reply:
The point wasn't actually about testing but about that I may in fact just want to run a given experiment both 1) as part of a master experiment and 2) as a standalone, possibly when I'm early in project iteration or want to use a cached version of a prior step. 

So, using my prior example, I often want to have a master experiment with ingredients 
[pull_data, fit_model, perform_tests], but have it be easy to run fit_model as a standalone experiment on cached data, rather than only being able to run it in the context of the full master experiment. It's actually quite convenient right now that I can use the exact same code both as its own standalone experiment and as part of a multi-part experiment, and I'd be sad to lose that for reasons other than just ease of unit testing (though it certainly is easier to iterate on both coding and testing separate components of experiments when they're runnable as their own experiments)

Reply:
The current proposal gives more flexibility than ingredients. Given the workflow you are describing, I think you'll have to code it differently, but it's still possible. You'll have   four experiments instead of one experiment and three ingredients.

>  I quite enjoy the current ingredient design insofar as an Ingredient being itself an Experiment 

That's not really true. An experiment is an ingredient, but an ingredient isn' t an experiment. The Experiment class is subclassing the ingredient class currently. 

I'm positive that your use case is supported given some code changes on your side. When we agree on a new API, I'll produce a Gist to implement your workflow with it and then you can tell me if you are satisfied with how it looks.